File: lab04/c_version/ex4.c
/*********************************************************************
* Title: CSCI203 - Assignment 1
* File: ex4.c - main() driver for exercise 4 implementation
* Author: Dinh Che (codeninja55) andrew at codeninja.me
* UOW Details: dbac496 | 5721970 | dbac496 at uowmail.edu.au
* Last modified: 2018.08.18
*********************************************************************/
#include <stdbool.h>
#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>
#include <memory.h>
#include <helper.h>

typedef struct Node {
    long int key;
    struct Node *left, *right;
    int height;
} Node;

/* GLOBAL VARS */
int COUNTER = 0;

/* FUNCTION PROTOTYPES */
Node* bst_insert_first(long int);
void bst_insert(Node*, long int);
void in_order_sort(Node*);
Node* avl_insert(Node*, long int);
Node* avl_rotate_right(Node*);
Node* avl_rotate_left(Node*);
Node* avl_double_rotate_right(Node*);
Node* avl_double_rotate_left(Node*);
int height(Node*);
int get_balance(Node*);
int max(int, int);

int main(int argc, const char* argv[])
{
    char filename[200];
    if (argc == 2) strcpy(filename, argv[1]);
    else {
        printf("[USAGE]: ex4 <filename> | OR \nInput file name: ");
        scanf("%s", filename);
    }

    FILE *fd;
    if ( (fd = fopen(filename, "r")) == NULL ) {
        fprintf(stderr, "[DEBUG]: Failed to open file [ %s ]", strerror(1));
        exit(1);
    }

    Node *ROOT = NULL;
    Node *AVL_ROOT = NULL;
    // AVL_ROOT = (Node *) ec_malloc(sizeof(Node));
    long int key;

    while ( fscanf(fd, "%li\n", &key) == 1 ) {
        // printf("[DEBUG]: Read: [ %li ]\n", key);
        // if ( ROOT == NULL ) ROOT = bst_insert_first(key);
        // else bst_insert(ROOT, key);
        AVL_ROOT = avl_insert(AVL_ROOT, key);
    }

    fclose(fd);
    // in_order_sort(ROOT);
    in_order_sort(AVL_ROOT);
    return 0;
}


// Insert the first node into the position of the ROOT pointer
Node* bst_insert_first(long int key)
{
    Node *first_node_ptr = (Node *) ec_malloc(sizeof(Node));
    first_node_ptr->key = key;
    first_node_ptr->height = 0;
    first_node_ptr->left = first_node_ptr->right = NULL;
    return first_node_ptr;
}

// Recursively start at the root node and find the position to insert the next node.
// Non-balanced binary search tree.
void bst_insert(Node* node, long int key)
{
    Node *next = NULL;
    bool left;

    if ( key == node->key )  // Already in the tree
        return;
    else if ( key < node->key ) {  // Go LEFT
        next = node->left;
        left = true;
    } else {  // Go RIGHT
        next = node->right;
        left = false;
    }

    if ( next != NULL ) bst_insert(next, key);  // Traverse down the tree
    else {
        Node *new_node_ptr = (Node *) ec_malloc(sizeof(Node));
        new_node_ptr->key = key;
        new_node_ptr->left = new_node_ptr->right = NULL;
        new_node_ptr->height = 0;
        if ( left ) node->left = new_node_ptr;
        else node->right = new_node_ptr;
    }
}

// Traverse and sort the BST in-order
void in_order_sort(struct Node* tree)
{
    if ( tree != NULL ) {
        if ( tree->left != NULL ) in_order_sort(tree->left);
        if ( COUNTER != 0 && (COUNTER % 10) == 0 ) printf("\n");
        COUNTER++;
        printf("%5ld", tree->key);
        if ( tree->right != NULL ) in_order_sort(tree->right);
    }
}

Node* avl_insert(Node* node, long int key)
{
    if ( node == NULL ) {
        Node *new_node = (Node *) ec_malloc(sizeof(Node));
        new_node->key = key;
        new_node->left = new_node->right = NULL;
        new_node->height = 0;
        return new_node;
    } else if ( key < node->key ) {
        node->left = avl_insert(node->left, key);
    } else if ( key > node->key ) {
        node->right = avl_insert(node->right, key);
    } else return node;  // Equal keys not allowed

    // tree->height = 1 + max((tree->left)->height, (tree->right)->height);
    node->height = 1 + max(height(node->left), height(node->right));

    int balance = get_balance(node);

    // Checking the imbalance and balancing the BST.
    /* CASE 1 - left left case */
    if ( balance > 1 && key < (node->left)->key )
        return avl_rotate_right(node);

    /* CASE 2 - right right case */
    if ( balance < -1 && key > (node->right)->key )
        return avl_rotate_left(node);

    /* CASE 3 - left right case */
    if ( balance > 1 && key > (node->left)->key )
        return avl_double_rotate_right(node);

    /* CASE 4 - right left case */
    if ( balance < -1 && key < (node->right)->key )
        return avl_double_rotate_left(node);

    return node;
}


Node* avl_rotate_right(Node* n2)
{
    Node *n1 = n2->left;
    Node *t2 = n1->right;

    n1->right = n2;
    n2->left = t2;
    n2->height = 1 + max(height(n2->left), height(n2->right));
    n1->height = 1 + max(height(n1->left), height(n2));
    return n1;
}

Node* avl_rotate_left(Node* n2)
{
    Node *n1 = n2->right;
    Node *t2 = n1->left;

    n1->left = n2;
    n2->right = t2;

    n2->height = 1 + max(height(n2->left), height(n2->right));
    n1->height = 1 + max(height(n2), height(n1->right));
    return n1;
}

Node* avl_double_rotate_right(Node* n3)
{
    n3->left = avl_rotate_left(n3->left);
    return avl_rotate_right(n3);
}

Node* avl_double_rotate_left(Node* n3)
{
    n3->right = avl_rotate_right(n3->right);
    return avl_rotate_left(n3);
}

/* HELPER FUNCTIONS */

// A helper function to get the height of the tree
int height(Node *n)
{
    return (n == NULL) ? 0 : n->height;
}

// Get Balance factor of node N
int get_balance(Node *n)
{
    return (n == NULL) ? 0 : height(n->left) - height(n->right);
}

int max(int a, int b)
{
    return (a > b) ? a : b;
}

File: lab04/py_version/ex4.py
#######################################################################
# Title: CSCI203 - Exercise 4
# File: ex4.py - main() for exercise 4 execution
# Author: Dinh Che (codeninja55) | andrew at codeninja55.me
# UOW Details: dbac496 | 5721970 | dbac496 at uowmail.edu.au
# Last modified: 2018.08.14
# REFERENCE: https://www.geeksforgeeks.org/avl-tree-set-1-insertion/
#######################################################################

import os
import sys
from argparse import ArgumentParser
from pathlib import Path

COUNTER = 0


class TreeNode(object):
    """
    Generic Tree Node class.
    """

    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None
        self.height = 1


class AVLTree(object):
    """
    The AVL Tree class which supports all AVL operations.
    """

    def get_height(self, node):
        return 0 if not node else node.height

    def get_balance(self, node):
        return 0 if not node else self.get_height(node.left) - self.get_height(node.right)

    def inorder_sort(self, root):
        global COUNTER
        if root.left:
            self.inorder_sort(root.left)

        if not (COUNTER == 0) and (COUNTER % 10) == 0:
            print('\n', end='')
        COUNTER += 1
        print("{0:5d}".format(root.key), end='')

        if root.right:
            self.inorder_sort(root.right)

    def right_rotate(self, node2):
        node1 = node2.left
        temp2 = node1.right
        # Perform rotation here
        node1.right = node2
        node2.left = temp2
        # Update the new heights
        node2.height = 1 + max(self.get_height(node2.left), self.get_height(node2.right))
        node1.height = 1 + max(self.get_height(node2), self.get_height(node1.right))
        return node1

    def left_rotate(self, node2):
        node1 = node2.right
        temp1 = node1.left
        # Perform rotation here
        node1.left = node2
        node2.right = temp1
        # Update heights
        node2.height = 1 + max(self.get_height(node2.left), self.get_height(node2.right))
        node1.height = 1 + max(self.get_height(node2), self.get_height(node1.right))
        return node1

    def double_right_rotate(self, node3):
        node3.left = self.left_rotate(node3.left)
        return self.right_rotate(node3)

    def double_left_rotate(self, node3):
        node3.right = self.right_rotate(node3.right)
        return self.left_rotate(node3)

    def insert(self, root, key):
        """
        A recursive class function to insert a new node into the tree with rotation balancing.
        :param root: The root node of the tree.
        :param key: The new key to insert.
        :return: The node inserted or the node after rotation.
        """
        # Perform normal BST
        # root = self.bst_insert(root, key)
        if not root:
            return TreeNode(key)
        elif key == root.key:
            return root
        elif key < root.key:
            root.left = self.insert(root.left, key)
        else:
            root.right = self.insert(root.right, key)

        # Update the height of the parent node
        root.height = 1 + max(self.get_height(root.left), self.get_height(root.right))

        # Get the balance factor to use in different cases
        balance = self.get_balance(root)

        # If the node is imbalanced, then try out each of the four cases
        # CASE 1
        if balance > 1 and key < root.left.key:
            return self.right_rotate(root)

        # CASE 2
        if balance < -1 and key > root.right.key:
            return self.left_rotate(root)

        # Case 3
        if balance > 1 and key > root.left.key:
            return self.double_right_rotate(root)

        # Case 4
        if balance < -1 and key < root.right.key:
            return self.double_left_rotate(root)

        # Otherwise if no balancing rotations needed, return the node
        return root


class BinarySearchTree(object):
    def bst_insert(self, root, key):
        """
        A recursive function to insert a new TreeNode into the tree without rotation balancing.
        :param root: The root node of the tree or None.
        :param key: The new key to insert.
        :return: The updated root node or a new TreeNode(key)
        """
        left = False

        if not root:  # No root node so create new one
            return TreeNode(key)

        if key == root.key:  # Already in the tree
            return None
        elif key < root.key:  # Go LEFT
            next_node = root.left
            left = True
        else:
            next_node = root.right
            left = False

        if next_node is not None:  # If next is not set, traverse the tree
            self.bst_insert(next_node, key)
        else:
            if left:
                root.left = TreeNode(key)
            else:
                root.right = TreeNode(key)
        return root


def read_file(filename):
    global AVL_TREE
    global ROOT

    if filename:
        with open(filename, 'r') as fin:
            if fin.closed:
                sys.exit(1)

            for line in fin:
                try:
                    key = int(line)
                    ROOT = AVL_TREE.insert(ROOT, key)
                except ValueError as e:
                    print('[DEBUG]: Cannot convert line to integer [ {} ]\n'
                          .format(e))


def parse_args():
    parser = ArgumentParser(prog='CN55 AVL Tree Implementation',
                            description='''This python script is an implementation
                            of CSCI203 - Algorithms and Data Structures (Spring 2018)
                            AVL Binary Search Trees.''')
    parser.add_argument('filename', help='Enter file name to read from.')
    parser.add_argument('-f', '--filename', required=False, dest='filename',
                        help='Enter file name to read from.')
    return parser.parse_args()


if __name__ == '__main__':
    AVL_TREE = AVLTree()
    ROOT = None

    print(os.getcwd())
    args = parse_args()
    if args.filename:
        file_path = Path(os.getcwd()) / args.filename
        read_file(file_path)
    AVL_TREE.inorder_sort(ROOT)
